File: Caronae/Falae/FalaeViewController.m
@import ActionSheetPicker_3_0;

#import <sys/utsname.h>
#import "CaronaeAlertController.h"
#import "FalaeViewController.h"
#import "Caronae-Swift.h"

/**
 * Returns the model of the current device.
 */
NSString *deviceName() {
    struct utsname systemInfo;
    uname(&systemInfo);
    
    return [NSString stringWithCString:systemInfo.machine encoding:NSUTF8StringEncoding];
}

@interface FalaeViewController ()

@property (nonatomic) NSString *selectedType;
@property (nonatomic) NSString *selectedTypeCute;
@property (nonatomic) int selectedTypeInitialIndex;
@property (nonatomic) NSArray *messageTypes;
@property (nonatomic) User *reportedUser;

@end

@implementation FalaeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _messageTypes = @[@"Sugestão", @"Denúncia", @"Dúvida"];
    
    if (_reportedUser) {
        _selectedType = @"report";
        _selectedTypeCute = @"Denúncia";
        _selectedTypeInitialIndex = (int)[_messageTypes indexOfObject:_selectedTypeCute];
        [_typeButton setTitle:@"Denúncia" forState:UIControlStateNormal];
        _subjectTextField.text = [NSString stringWithFormat:@"Denúncia sobre usuário %@ (id: %ld)", _reportedUser.name, (long)_reportedUser.id];
        _subjectTextField.enabled = NO;
    }
    else {
        _selectedType = @"complaint";
        _selectedTypeCute = @"Denúncia";
        _selectedTypeInitialIndex = (int)[_messageTypes indexOfObject:_selectedTypeCute];
    }
}

- (void)setReport:(User *)user {
    _reportedUser = user;
}

- (void)sendMessage:(NSDictionary *)message {
    
    [self showLoadingHUD:YES];
    [CaronaeAPIHTTPSessionManager.instance POST:@"/falae/sendMessage" parameters:message success:^(NSURLSessionDataTask * _Nonnull task, id  _Nonnull responseObject) {
        [self showLoadingHUD:NO];
        
        [CaronaeAlertController presentOkAlertWithTitle:@"Mensagem enviada!" message:@"Obrigado por nos mandar uma mensagem. Nossa equipe irá entrar em contato em breve." handler:^{
            [self.navigationController popViewControllerAnimated:YES];
        }];
        
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        [self showLoadingHUD:NO];
        NSLog(@"Error: %@", error.localizedDescription);
        
        [CaronaeAlertController presentOkAlertWithTitle:@"Mensagem não enviada" message:@"Ocorreu um erro enviando sua mensagem. Verifique sua conexão e tente novamente."];
    }];
}

#pragma mark - IBActions

- (IBAction)didTapSelectTypeButton:(id)sender {
    [self.view endEditing:YES];
    [ActionSheetStringPicker showPickerWithTitle:@"Qual o motivo do seu contato?"
                                            rows:_messageTypes                                                          initialSelection:_selectedTypeInitialIndex
                                       doneBlock:^(ActionSheetStringPicker *picker, NSInteger selectedIndex, id selectedValue) {
                                           _selectedTypeCute = selectedValue;
                                           if ([selectedValue isEqualToString:@"Sugestão"]) {
                                               _selectedType = @"suggestion";
                                           }
                                           else if ([selectedValue isEqualToString:@"Denúncia"]) {
                                               _selectedType = @"report";
                                           }
                                           else if ([selectedValue isEqualToString:@"Dúvida"]) {
                                               _selectedType = @"help";
                                           }
                                           else {
                                               _selectedType = @"other";
                                           }
                                           [_typeButton setTitle:selectedValue forState:UIControlStateNormal];
                                       }
                                     cancelBlock:nil origin:sender];
}

- (IBAction)didTapSendButton:(id)sender {
    [self.view endEditing:YES];
    
    NSString *messageText = [_messageTextView.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if (messageText.length == 0) {
        [CaronaeAlertController presentOkAlertWithTitle:@"Ops!" message:@"Parece que você esqueceu de preencher sua mensagem."];
        return;
    }
    
    NSString *type = _selectedType;
    NSString *subject = [NSString stringWithFormat:@"[%@] %@", _selectedTypeCute, _subjectTextField.text];
    
    NSString *appVersionString = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
    NSString *appBuildString = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleVersion"];
    NSString *versionBuildString = [NSString stringWithFormat:@"%@ (build %@)", appVersionString, appBuildString];
    NSString *osVersion = [[UIDevice currentDevice] systemVersion];
    NSString *device = deviceName();
    
    NSString *text = [NSString stringWithFormat:@"%@\n\n--------------------------------\nDevice: %@ (iOS %@)\nVersão do app: %@", messageText, device, osVersion, versionBuildString];
    
    NSDictionary *message = @{@"type": type,
                              @"subject": subject,
                              @"message": text};
    
    [self sendMessage:message];
}


#pragma mark - Etc

- (void)showLoadingHUD:(BOOL)loading {
    self.navigationItem.rightBarButtonItem = loading ? self.loadingButton : self.sendButton;
}

- (UIBarButtonItem *)loadingButton {
    if (!_loadingButton) {
        UIActivityIndicatorView *spinner = [[UIActivityIndicatorView alloc]
                                            initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
        [spinner startAnimating];
        _loadingButton = [[UIBarButtonItem alloc] initWithCustomView:spinner];
    }
    return _loadingButton;
}

@end

File: Caronae/Falae/FalaeViewController.swift
import UIKit
import ActionSheetPicker_3_0
import SVProgressHUD

class FalaeViewController: UIViewController {
    @IBOutlet weak var typeButton: UIButton!
    @IBOutlet weak var subjectTextField: UITextField!
    @IBOutlet weak var messageTextView: UITextView!
    @IBOutlet weak var sendButton: UIBarButtonItem!
    
    var reportedUser: User?
    let messageTypes = ["Reclamação", "Sugestão", "Denúncia", "Dúvida"]
    var selectedTypeIndex = 0
    var selectedType = "Reclamação" {
        didSet {
            typeButton.setTitle(selectedType, for: .normal)
            self.selectedTypeIndex = messageTypes.index(of: selectedType)!
        }
    }
    
    var deviceName: String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let reportedUser = reportedUser {
            selectedType = "Denúncia"
            typeButton.isEnabled = false
            subjectTextField.text = String(format: "Denúncia sobre usuário %@ (id: %ld)", reportedUser.name, reportedUser.id)
            subjectTextField.isEnabled = false
        }
    }
    
    func setReport(user: User) {
        reportedUser = user
    }
    
    
    // MARK: IBActions
    
    @IBAction func didTapSelectTypeButton(_ sender: Any) {
        view.endEditing(true)
        ActionSheetStringPicker.show(withTitle: "Qual o motivo do seu contato?",
                                     rows: messageTypes, initialSelection: selectedTypeIndex,
                                     doneBlock: { _, _, selectedValue in
                                        self.selectedType = selectedValue as! String
                                     }, cancel: nil, origin: sender)
    }
    
    @IBAction func didTapSendButton(_ sender: Any) {
        view.endEditing(true)
        
        let messageText = messageTextView.text.trimmingCharacters(in: .whitespacesAndNewlines)
        if messageText.isEmpty {
            CaronaeAlertController.presentOkAlert(withTitle: "Ops!", message: "Parece que você esqueceu de preencher sua mensagem.")
            return
        }
        let subject = String(format: "[%@] %@", selectedType, subjectTextField.text!)
        
        let appVersionString = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as! String
        let appBuildString = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as! String
        let versionBuildString = String(format: "%@ (build %@)", appVersionString, appBuildString)
        let osVersion = UIDevice.current.systemVersion
        let device = deviceName
        let message = String(format: "%@\n\n--------------------------------\nDevice: %@ (iOS %@)\nVersão do app: %@", messageText, device, osVersion, versionBuildString)
        
        sendButton.isEnabled = false
        SVProgressHUD.show()
        FalaeService.instance.sendMessage(subject: subject, message: message, success: {
            SVProgressHUD.dismiss()
            CaronaeAlertController.presentOkAlert(withTitle: "Mensagem enviada!", message: "Obrigado por nos mandar uma mensagem. Nossa equipe irá entrar em contato em breve.", handler: {
                self.navigationController?.popViewController(animated: true)
            })
        }) { error in
            self.sendButton.isEnabled = true
            SVProgressHUD.dismiss()
            NSLog("Error: %@", error!.localizedDescription)
            CaronaeAlertController.presentOkAlert(withTitle: "Mensagem não enviada", message: "Ocorreu um erro enviando sua mensagem. Verifique sua conexão e tente novamente.")
        }
    }
}
